File: coauthor/latex/tikzpicture.py
import re
import os

from ..logger import logger

from ..utils.file import read_file, write_file
from ..utils.prompt import render_prompt

from .tex_tools import compile_latex_to_pdf


# maybe in the future move to a separate file in the agent path
TIKZ_TEMPLATE = r"""
\documentclass[tikz,border=10pt]{standalone}
\usepackage{tikz}
\usepackage{pgfplots}
\usetikzlibrary{positioning}
\usetikzlibrary{patterns}
\usetikzlibrary{arrows.meta, shapes.geometric, matrix, calc, decorations.pathreplacing}
\usetikzlibrary{shapes, arrows}

\begin{document}
{{ tikzpicture }}
\end{document}
"""


def extract_tikzpictures_with_labels(latexFile: str) -> list[tuple[str, list[str]]]:
    """Extract TikZ pictures and their associated labels from LaTeX file, returns list of (label, pictures) tuples."""
    content = read_file(latexFile)

    # Regular expression to match entire figure environments with labels and tikzpicture environments
    figure_pattern = re.compile(r"(\\begin{figure}.*?\\label\{.*?\}.*?\\end{figure})", re.DOTALL)
    tikz_pattern = re.compile(r"\\begin{tikzpicture}.*?\\end{tikzpicture}", re.DOTALL)

    labeled_tikzpictures = []
    for figure in figure_pattern.findall(content):
        label_match = re.search(r"\\label\{(.*?)\}", figure)
        if label_match:
            label = label_match.group(1)
            tikz_matches = tikz_pattern.findall(figure)
            if tikz_matches:
                labeled_tikzpictures.append((label, tikz_matches))

    return labeled_tikzpictures


def extract_and_compile_tikzpictures_with_labels(latexFile: str) -> list[str]:
    """Extract and compile TikZ pictures from LaTeX file into standalone PDFs, returns PDF paths."""
    input_dir = os.path.dirname(latexFile)
    input_name = os.path.splitext(os.path.basename(latexFile))[0]
    build_dir = os.path.join(input_dir, "build", input_name)
    os.makedirs(build_dir, exist_ok=True)

    # Extract TikZ pictures
    logger.debug("Extracting TikZ pictures with labels...")
    labeled_tikzpictures = extract_tikzpictures_with_labels(latexFile)
    logger.debug(f"Found {len(labeled_tikzpictures)} labeled TikZ pictures.")

    compiled_files = []
    for label, tikzpictures in labeled_tikzpictures:
        # Handle multiple TikZ pictures with same label by adding a,b,c suffixes
        suffixes = [chr(97 + i) if len(tikzpictures) > 1 else None for i in range(len(tikzpictures))]

        for tikzpicture, suffix in zip(tikzpictures, suffixes):
            # Create and compile standalone LaTeX file
            tex_file = create_standalone_latex_with_labels(tikzpicture, label, build_dir, suffix)
            compile_latex_to_pdf(tex_file)
            pdf_file = f"{os.path.splitext(tex_file)[0]}.pdf"
            compiled_files.append(pdf_file)

            # Clean up auxiliary files
            for aux_file in [f"{os.path.splitext(tex_file)[0]}{ext}" for ext in (".aux", ".log")]:
                if os.path.exists(aux_file):
                    os.remove(aux_file)

    return compiled_files


def create_standalone_latex_with_labels(tikzpicture: str, label: str, build_dir: str, suffix: str | None = None) -> str:
    """Create a standalone LaTeX file containing a TikZ picture with labels."""
    standalone_content = render_prompt(TIKZ_TEMPLATE, {"tikzpicture": tikzpicture})
    filename = os.path.join(build_dir, f"{label}_{suffix}.tex" if suffix else f"{label}.tex")
    write_file(filename, standalone_content)

    return filename

File: coauthor-for-vs-code/src/latex/tikzPicture.ts
import * as path from 'path';
import { debug, error, initializeLogging } from '../logger/logUtils';
import {
  readFile,
  fileExists,
  getWorkspacePath,
  writeFile,
  createDirectory,
} from '../utils/fileUtils';
import { compileLatexToPdf } from './texTools';
import * as nunjucks from 'nunjucks';
import { renderPrompt } from '../utils/promptUtils';

const CHANNEL = 'LaTeX';
initializeLogging(CHANNEL);

// Configure nunjucks
nunjucks.configure({ autoescape: false });

// Add the TikZ template
// maybe in the future move to a separate file explorer.path
const TIKZ_TEMPLATE = `
\\documentclass[tikz,border=10pt]{standalone}
\\usepackage{tikz}
\\usepackage{pgfplots}
\\usetikzlibrary{positioning}
\\usetikzlibrary{patterns}
\\usetikzlibrary{arrows.meta, shapes.geometric, matrix, calc, decorations.pathreplacing}
\\usetikzlibrary{shapes, arrows}

\\begin{document}
{{ tikzpicture }}
\\end{document}
`;

/**
 * Extract TikZ pictures with their labels from a LaTeX file
 * @param latexFile Path to the LaTeX file
 * @returns Array of [label, tikzpictures] tuples
 */
export async function extractTikzPicturesWithLabels(
  latexFile: string,
): Promise<[string, string[]][]> {
  try {
    const content = await readFile(latexFile);

    // Regular expressions to match figure environments and tikzpictures
    const figurePattern =
      /\\begin{figure}.*?\\label\{(.*?)\}.*?\\end{figure}/gs;
    const tikzPattern = /\\begin{tikzpicture}.*?\\end{tikzpicture}/gs;

    const labeledTikzPictures: [string, string[]][] = [];

    let figureMatch;
    while ((figureMatch = figurePattern.exec(content)) !== null) {
      const figureContent = figureMatch[0];
      const label = figureMatch[1];

      // Find all tikzpictures in this figure
      const tikzMatches = [...figureContent.matchAll(tikzPattern)].map(
        (match) => match[0],
      );

      if (tikzMatches.length > 0) {
        labeledTikzPictures.push([label, tikzMatches]);
        debug(CHANNEL, `Found TikZ picture with label: ${label}`);
      }
    }

    return labeledTikzPictures;
  } catch (err) {
    error(
      CHANNEL,
      `Error extracting TikZ pictures: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}

/**
 * Create a standalone LaTeX file for a TikZ picture
 * @param tikzPicture TikZ picture content
 * @param label Label for the figure
 * @param buildDir Build directory path
 * @param suffix Optional suffix for multiple pictures with same label
 * @returns Path to created LaTeX file
 */
export async function createStandaloneLatexWithLabels(
  tikzPicture: string,
  label: string,
  buildDir: string,
  suffix?: string,
): Promise<string> {
  try {
    // Use renderPrompt instead of nunjucks directly
    const standaloneContent = await renderPrompt(TIKZ_TEMPLATE, {
      tikzpicture: tikzPicture,
    });

    // Create filename
    const filename = suffix ? `${label}_${suffix}.tex` : `${label}.tex`;
    const filePath = path.join(buildDir, filename);

    // Write file
    await writeFile(filePath, standaloneContent);
    debug(CHANNEL, `Created standalone LaTeX file: ${filePath}`);

    return filePath;
  } catch (err) {
    error(
      CHANNEL,
      `Error creating standalone LaTeX: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}

/**
 * Extract and compile TikZ pictures from a LaTeX file
 * @param latexFile Path to the LaTeX file
 * @returns Array of paths to compiled PDF files
 */
export async function extractAndCompileTikzPicturesWithLabels(
  latexFile: string,
): Promise<string[]> {
  try {
    // Setup build directory
    const inputDir = path.dirname(latexFile);
    const inputName = path.parse(path.basename(latexFile)).name;
    const buildDir = path.join(inputDir, 'build', inputName);
    await createDirectory(buildDir);

    debug(CHANNEL, `Extracting TikZ pictures from ${latexFile}`);
    const labeledTikzPictures = await extractTikzPicturesWithLabels(latexFile);
    debug(CHANNEL, `Found ${labeledTikzPictures.length} labeled TikZ pictures`);

    const compiledFiles: string[] = [];

    for (const [label, tikzPictures] of labeledTikzPictures) {
      // Generate suffixes for multiple pictures with same label
      const suffixes =
        tikzPictures.length > 1
          ? tikzPictures.map((_, i) => String.fromCharCode(97 + i)) // a, b, c, ...
          : [undefined];

      for (let i = 0; i < tikzPictures.length; i++) {
        const tikzPicture = tikzPictures[i];
        const suffix = suffixes[i];

        // Create and compile standalone LaTeX file
        const texFile = await createStandaloneLatexWithLabels(
          tikzPicture,
          label,
          buildDir,
          suffix,
        );
        await compileLatexToPdf(texFile);

        const pdfFile = texFile.replace(/\.tex$/, '.pdf');
        if (await fileExists(pdfFile)) {
          compiledFiles.push(pdfFile);
          debug(CHANNEL, `Successfully compiled: ${pdfFile}`);
        }
      }
    }

    return compiledFiles;
  } catch (err) {
    error(
      CHANNEL,
      `Error extracting and compiling TikZ pictures: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}
